crates/dbsp/src/operator/dynamic/balance/balancer.rs

```diff
@@ -0,0 +1,832 @@
+        i2: VariableIndex,
+        affected_variables: &mut BTreeSet<VariableIndex>,
+    ) {
+        let v1 = &maxsat.variables()[i1.0].domain;
+        //println!("current variables: {:?}", maxsat.variables());
+
+        // If the left domain doesn't contain Shard or Broadcast, then the right domain
+        // cannot contain Shard.
+        let remove_shard = !v1.contains_key(&PartitioningPolicy::Shard.into())
+            && !v1.contains_key(&PartitioningPolicy::Broadcast.into());
+
+        // If the left domain doesn't contain Shard or Balance, then the right domain
+        // cannot contain Broadcast.
+        let remove_broadcast = !v1.contains_key(&PartitioningPolicy::Shard.into())
+            && !v1.contains_key(&PartitioningPolicy::Balance.into());
+
+        // If the left domain doesn't contain Broadcast, then the right domain
+        // cannot contain Balance.
+        let remove_balance = !v1.contains_key(&PartitioningPolicy::Broadcast.into());
+
+        let mut v2_modified = false;
+
+        if remove_shard {
+            // println!("removing SHARD from {}", maxsat.variables()[i2.0].name);
+            v2_modified |= maxsat.remove_variable_assignment(i2, PartitioningPolicy::Shard.into());
+        }
+        if remove_broadcast {
+            // println!("removing BROADCAST from {}", maxsat.variables()[i2.0].name);
+            v2_modified |=
+                maxsat.remove_variable_assignment(i2, PartitioningPolicy::Broadcast.into());
+        }
+        if remove_balance {
+            // println!("removing BALANCE from {}", maxsat.variables()[i2.0].name);
+            v2_modified |=
+                maxsat.remove_variable_assignment(i2, PartitioningPolicy::Balance.into());
+        }
+
+        if v2_modified {
+            affected_variables.insert(i2);
+        }
+    }
+}
+
+impl HardConstraint for JoinConstraint {
+    fn propagate(&self, maxsat: &mut MaxSat, affected_variables: &mut BTreeSet<VariableIndex>) {
+        if self.is_left_join {
+            //let v1 = &mut all_variables[self.v1.0].domain;
+            //if v1.remove(&PartitioningPolicy::Broadcast.into()).is_some() {
+            if maxsat.remove_variable_assignment(self.v1, PartitioningPolicy::Broadcast.into()) {
+                affected_variables.insert(self.v1);
+            }
+        }
+        self.propagate_inner(maxsat, self.v1, self.v2, affected_variables);
+        self.propagate_inner(maxsat, self.v2, self.v1, affected_variables);
+    }
+
+    fn variables(&self) -> Vec<VariableIndex> {
+        vec![self.v1, self.v2]
+    }
+}
+
+/// Partitioning policies supported by the balancer.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum PartitioningPolicy {
+    /// Shard the stream based on the hash of the key.
+    Shard = 0,
+    /// Broadcast the stream to all workers.
+    Broadcast = 1,
+    /// Split the stream into near-equal-sized partitions across all workers
```

Comment:
you don't know that they are near-equal, it really depends on the data distribution
you just know that you are hashing on the entire record
on the other hand, the fact that there are no duplicates in zsets should help